cli/src/commands/sign.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2023 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use clap_complete::ArgValueCandidates;
+use itertools::Itertools;
+use jj_lib::commit::CommitIteratorExt;
+use jj_lib::signing::SignBehavior;
+
+use crate::cli_util::CommandHelper;
+use crate::cli_util::RevisionArg;
+use crate::command_error::user_error;
+use crate::command_error::CommandError;
+use crate::complete;
+use crate::ui::Ui;
+
+/// Cryptographically sign a revision
+#[derive(clap::Args, Clone, Debug)]
+pub struct SignArgs {
+    /// What key to use, depends on the configured signing backend.
+    #[arg()]
+    key: Option<String>,
+    /// What revision(s) to sign
+    #[arg(
+        long, short,
+        default_value = "@",
+        value_name = "REVSETS",
+        add = ArgValueCandidates::new(complete::all_revisions),
+    )]
+    revisions: Vec<RevisionArg>,
+    /// Sign a commit that is not authored by you or was already signed.
+    #[arg(long, short)]
+    force: bool,
+    /// Drop the signature, explicitly "un-signing" the commit.
+    #[arg(long, short = 'D', conflicts_with = "force")]
+    drop: bool,
+}
+
+pub fn cmd_sign(ui: &mut Ui, command: &CommandHelper, args: &SignArgs) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+
+    let commits = workspace_command
+        .resolve_some_revsets_default_single(ui, &args.revisions)?
+        .into_iter()
+        .collect_vec();
+
+    let commit_ids = commits.iter().ids().collect_vec();
+    workspace_command.check_rewritable(commit_ids)?;
+
+    for commit in &commits {
+        if !args.force {
+            if !args.drop && commit.is_signed() {
+                return Err(user_error(
+                    "Commit is already signed, use --force to sign anyway",
+                ));
+            }
+            if commit.author().email != command.settings().user_email() {
+                return Err(user_error(
+                    "Commit is not authored by you, use --force to sign anyway",
+                ));
+            }
```

Comment:
Now that this version can work with multiple commits, those checks should show the offending one.
Or one***s***, actually, so they should eagerly collect lists of bad commits and print them all at once.

Reply:
nit: Also `if args.force` can be moved outside of the loop